index.test.js

```diff
@@ -15,15 +18,76 @@ describe("the JS entrypoint to the GitHub action", () => {
 
     const http_mock = new MockAdapter(axios);
     http_mock.onPost("https://example.com/harness/webhook").reply(200, {
-      api_url: "https://example.org/api",
-      harness_url: "https://example.org/harness/execution",
+      requestId: "req123",
       status: "RUNNING",
+      error: null,
+      uiUrl:
+        "app.harness.io/#/account/aaa/app/bbb/pipeline-execution/ccc/workflow-execution/undefined/details",
+      apiUrl:
+        "app.harness.io/api/external/v1/executions/aaa/status?accountId=bbb&appId=ccc",
+      message: null,
     });
 
-    const entrypoint = require(".");
+    const entrypoint = require("./index.js");
 
-    expect(consoleSpy).toHaveBeenCalledWith(
+    expect(coreSpy).toHaveBeenCalledWith(
       "Deploying application:pincushion (pin, cush ,ion) at v0"
     );
   });
+
+  test("runs the action and waits for deploy", () => {
+    // expect.assertions(1);
+
+    const coreSpy = jest.spyOn(core, "info");
+
+    process.env["INPUT_APPLICATION"] = "pincushion";
+    process.env["INPUT_SERVICES"] = "pin, cush ,ion";
+    process.env["INPUT_WEBHOOKURL"] = "https://example.com/harness/webhook";
+    process.env["INPUT_VERSION"] = "v0";
+    process.env["INPUT_WAITFORDEPLOY"] = "true";
+
+    const http_mock = new MockAdapter(axios);
+    http_mock.onPost("https://example.com/harness/webhook").reply(200, {
+      requestId: "req123",
+      status: "RUNNING",
+      error: null,
+      uiUrl:
+        "app.harness.io/#/account/aaa/app/bbb/pipeline-execution/ccc/workflow-execution/undefined/details",
+      apiUrl:
+        "app.harness.io/api/external/v1/executions/aaa/status?accountId=bbb&appId=ccc",
+      message: null,
+    });
+
+    http_mock
+      .onGet(
+        "app.harness.io/api/external/v1/executions/aaa/status?accountId=bbb&appId=ccc"
+      )
+      .replyOnce(200, {
+        status: "RUNNING",
+      })
+      .onGet(
+        "app.harness.io/api/external/v1/executions/aaa/status?accountId=bbb&appId=ccc"
+      )
+      .replyOnce(200, {
+        status: "RUNNING",
+      })
+      .onGet(
+        "app.harness.io/api/external/v1/executions/aaa/status?accountId=bbb&appId=ccc"
+      )
+      .replyOnce(200, {
+        status: "SUCCESS",
+      });
+
+    const entrypoint = require("./index.js");
+
+    expect(coreSpy.mock.calls).toEqual([
+      ["Deploying application:pincushion (pin, cush ,ion) at v0"],
+      ["Sending request to start deployment"],
+      ["Checking response from request to start deployment"],
+      ["🚀 Deployment pipeline is now running on Harness"],
+      [
+        "Harness deploy submitted, view at app.harness.io/#/account/aaa/app/bbb/pipeline-execution/ccc/workflow-execution/undefined/details",
+      ],
```

Comment:
This test isn't ideal because it's too specific, but also because it doesn't actually test what we need it to here - that setting the wait for deploy option does something different. I'm pretty sure it's because the spy isn't the same instance as the one that is used by the watchDeployment function.
Leaving it here intentionally to follow back up on as a PR on it's own, rather than roll it into the wider PR here.